### PR TITLE
ci: fix build issue on m1

### PR DESCRIPTION
**Pull request by zirain**
Commit Message: fix build issue on m1
Additional Description: related to https://gitlab.freedesktop.org/pkg-config/pkg-config/-/issues/81
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #33225
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by zirain** [comment]
@phlax I'm not sure this's the right fix, just raise up for someone know this better than me.

File: bazel/rules_foreign_cc.patch
Comment (phlax) [inline bazel/rules_foreign_cc.patch:10]: wondering if we only want these flags for m1 where its required

Comment (zirain) [reply bazel/rules_foreign_cc.patch:10]: can we use `select` here?

Comment (phlax) [reply bazel/rules_foreign_cc.patch:10]: i think so - not sure how that is specified in terms of platform/arch for m1

Comment (phlax) [reply bazel/rules_foreign_cc.patch:10]: ... i guess arch=arm and platform = macos

Comment (phlax) [reply bazel/rules_foreign_cc.patch:10]: nm - i see https://github.com/envoyproxy/envoy/blob/4a3989af957c9b99683c142bb72a6b4f467b667e/bazel/BUILD#L854-L861

**Review by phlax — COMMENTED** [review]
im thinking this patch should probably go upstream - altho if it fixes the m1 build for now i dont want to block

perhaps add a comment above the diff linking the issue it resolves - otherwise it may get removed (as it wouldnt break ci) or left after it was really needed

File: bazel/rules_foreign_cc.patch
Comment (phlax) [reply bazel/rules_foreign_cc.patch:10]: > ... here?

also just realized this is a rule so no wrt to `select` no you cant here

not sure exactly how you handle platform detection in a rule

Comment (phlax) [reply bazel/rules_foreign_cc.patch:10]: https://bazel.build/rules/lib/builtins/ctx#target_platform_has_constraint

```starlark
    if ctx.target_platform_has_constraint(os="mac") and ctx.target_platform_has_constraint(cpu="arm64"):
        args = ["--mac", "--args"]
   else:
       args = []
   args += ["remaining", "args"]
```

**Comment by zirain** [comment]
> im thinking this patch should probably go upstream - altho if it fixes the m1 build for now i dont want to block
> 
> perhaps add a comment above the diff linking the issue it resolves - otherwise it may get removed (as it wouldnt break ci) or left after it was really needed

address with comment, and open a ticket here: https://github.com/bazelbuild/rules_foreign_cc/issues/1186

File: bazel/rules_foreign_cc.patch
Comment (phlax) [reply bazel/rules_foreign_cc.patch:10]: @zirain if you can address this part im happy to land

i would rather not disable `no-int-conversion` for all platforms - and would need a second opinion before landing

File: bazel/rules_foreign_cc.patch
Comment (phlax) [inline bazel/rules_foreign_cc.patch:20]:
```suggestion
+    if os_name(ctx) == "macos" and ctx.platform_has_constraint(cpu = "arm64"):
```

im thinking this doesnt matter _that_ much as its just for compiling pkg_config afaict but im also thinking the patch is more likely to be accepted upstream if its targetted to need

File: bazel/rules_foreign_cc.patch
Comment (zirain) [reply bazel/rules_foreign_cc.patch]: update, it's needed on mac intel with latest xcode.

BTW, need merge https://github.com/envoyproxy/envoy/pull/33231 first?

Comment (phlax) [reply bazel/rules_foreign_cc.patch]: ah, i see, thanks for clarifying

File: bazel/rules_foreign_cc.patch
Comment (zirain) [reply bazel/rules_foreign_cc.patch:20]: same issue on intel mac with xcode 15.3

**Review by phlax — DISMISSED** [review]
lgtm, thanks for resolving this @zirain

**Comment by phlax** [comment]
dont worry its not `required` and retesting will just retest the same (merge) commit

**Review by phlax — APPROVED** [review]

**Comment by zirain** [comment]
FYI: send a [PR](https://github.com/bazelbuild/rules_foreign_cc/pull/1187) to upstream, will remove this after it's merged.

**Pull request merged**
